bluebutton-data-pipeline-fhir-load/src/main/java/gov/hhs/cms/bluebutton/datapipeline/fhir/transform/DataTransformer.java

```diff
@@ -929,6 +948,85 @@ private TransformedBundle transformPartDEvent(RifRecordEvent<PartDEventRow> rifR
 	}
 
 	/**
+	 * @param rifRecordEvent
+	 *            the source {@link RifRecordEvent} to be transformed
+	 * @return the {@link TransformedBundle} that is the result of transforming
+	 *         the specified {@link RifRecordEvent}
+	 */
+	private TransformedBundle transformCarrierClaim(RifRecordEvent<CarrierClaimGroup> rifRecordEvent) {
+		if (rifRecordEvent == null)
+			throw new IllegalArgumentException();
+		CarrierClaimGroup claimGroup = rifRecordEvent.getRecord();
+		if (1 != claimGroup.version)
+			throw new IllegalArgumentException("Unsupported record version: " + claimGroup.version);
+		if (claimGroup.recordAction != RecordAction.INSERT)
+			// Will need refactoring to support other ops.
+			throw new BadCodeMonkeyException();
+
+		Bundle bundle = new Bundle();
+
+		ExplanationOfBenefit eob = new ExplanationOfBenefit();
+		eob.setId("ExplanationOfBenefit/" + claimGroup.claimId);
+		eob.getCoverage().setCoverage(referenceCoverage(claimGroup.beneficiaryId, COVERAGE_PLAN_PART_B));
+		eob.setPatient(referencePatient(claimGroup.beneficiaryId));
+
+		setPeriodStart(eob.getBillablePeriod(), claimGroup.dateFrom);
+		setPeriodEnd(eob.getBillablePeriod(), claimGroup.dateThrough);
+
+		/*
+		 * Referrals are represented as contained resources, because otherwise
+		 * updating them would require an extra roundtrip to the server (can't
+		 * think of an intelligent client-specified ID for them).
+		 */
+		Practitioner referrer = new Practitioner();
+		referrer.addIdentifier().setSystem(CODING_SYSTEM_NPI_US).setValue(claimGroup.referringPhysicianNpi);
+		Reference referrerRef = referencePractitioner(claimGroup.referringPhysicianNpi);
+		upsert(bundle, referrer, referrerRef.getReference());
+		ReferralRequest referral = new ReferralRequest();
```

Comment:
Thanks for reading so closely!

It actually should, but in a slightly different way: this should result in the resource getting created as a "contained resource," rather than a typical independent one. This specific mechanism in HAPI of creating a contained resource is mentioned by James here: https://github.com/jamesagnew/hapi-fhir/issues/113

Why a contained resource? Because the ReferralRequest should have the exact same lifetime as the EOB, If the EOB is deleted, it should be. If the claim is updated to blank out the `claimGroup.referringPhysicianNpi` field, the ReferralRequest should also be deleted. If it were an independent resource, I can't think of any way to accomplish that.

And this is me getting a bit ahead of myself: we're explicitly **not** handling updates or deletes yet. But I couldn't stop my brain from working through this particular edge case, so I included the solution when I found it.

But your question does highlight the fact that this choice and behavior is non-obvious, and absolutely should have a comment at least pointing it out. I'll add that now, and I'll file a story for later to start documenting our design choices around these kinds of problems.

Make sense?